component/storage/indexeddb/indexeddb.go

```diff
@@ -172,19 +247,78 @@ func (s *store) Get(key string) ([]byte, error) {
 	return []byte(data.Get("value").String()), nil
 }
 
-// Currently not implemented (TODO #2528).
 func (s *store) GetTags(key string) ([]storage.Tag, error) {
-	return nil, errors.New("not implemented")
+	if key == "" {
+		return nil, errors.New("key is mandatory")
+	}
+
+	req := s.db.Call("transaction", s.name).Call("objectStore", s.name).Call("get", key)
+
+	data, err := getResult(req)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get data: %w", err)
+	}
+
+	if !data.Truthy() {
+		return nil, storage.ErrDataNotFound
+	}
+
+	tagsBytes := []byte(data.Get("tags").String())
+
+	var tags []storage.Tag
+
+	err = json.Unmarshal(tagsBytes, &tags)
+	if err != nil {
+		return nil, fmt.Errorf("failed to unmarshal tags bytes: %w", err)
+	}
+
+	return tags, nil
 }
 
 // Currently not implemented (TODO #2528).
 func (s *store) GetBulk(keys ...string) ([][]byte, error) {
 	return nil, errors.New("not implemented")
 }
 
-// Currently not implemented (TODO #2528).
 func (s *store) Query(expression string, options ...storage.QueryOption) (storage.Iterator, error) {
-	return nil, errors.New("not implemented")
+	if expression == "" {
+		return nil, errInvalidQueryExpressionFormat
+	}
+
+	tagMapBytes, err := s.Get(tagMapKey)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get tag map: %w", err)
+	}
+
+	var tagMap tagMapping
+
+	err = json.Unmarshal(tagMapBytes, &tagMap)
+	if err != nil {
+		return nil, fmt.Errorf("failed to unmarshal tag map bytes: %w", err)
+	}
+
+	expressionSplit := strings.Split(expression, ":")
+	switch len(expressionSplit) {
+	case expressionTagNameOnlyLength:
+		expressionTagName := expressionSplit[0]
+
+		matchingDatabaseKeys := getDatabaseKeysMatchingTagName(tagMap, expressionTagName)
+
+		return &iterator{keys: matchingDatabaseKeys, store: s}, nil
+	case expressionTagNameAndValueLength:
+		expressionTagName := expressionSplit[0]
+		expressionTagValue := expressionSplit[1]
+
+		matchingDatabaseKeys, err :=
+			s.getDatabaseKeysMatchingTagNameAndValue(tagMap, expressionTagName, expressionTagValue)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get database keys matching tag name and value: %w", err)
+		}
+
+		return &iterator{keys: matchingDatabaseKeys, store: s}, nil
+	default:
+		return nil, errInvalidQueryExpressionFormat
```

Comment:
can you wrap this error and include the offending expression?

Reply:
Done!